eng/code-quality-reports/src/main/java/com/azure/tools/checkstyle/checks/AssignedOnceVariableToBeFinalCheck.java

```diff
@@ -0,0 +1,267 @@
+            case TokenTypes.BOR_ASSIGN:
+            case TokenTypes.BSR_ASSIGN:
+            case TokenTypes.BXOR_ASSIGN:
+            case TokenTypes.DIV_ASSIGN:
+            case TokenTypes.MINUS_ASSIGN:
+            case TokenTypes.MOD_ASSIGN:
+            case TokenTypes.SL_ASSIGN:
+            case TokenTypes.SR_ASSIGN:
+            case TokenTypes.STAR_ASSIGN:
+            case TokenTypes.INC:
+            case TokenTypes.POST_INC:
+            case TokenTypes.DEC:
+            case TokenTypes.POST_DEC:
+                checkAssignation(token);
+                break;
+            case TokenTypes.METHOD_DEF:
+            case TokenTypes.CTOR_DEF:
+                scopeParent = token;
+            default:
+                // Checkstyle complains if there's no default block in switch
+                break;
+        }
+    }
+
+    @Override
+    public void leaveToken(DetailAST token) {
+        switch (token.getType()) {
+            case TokenTypes.METHOD_DEF:
+            case TokenTypes.CTOR_DEF:
+                scopeParent = null;
+                currentScopeParameterSet = null;
+            default:
+                break;
+        }
+    }
+
+    @Override
+    public void finishTree(DetailAST token) {
+        for (DetailAST field : nonFinalFields) {
+            final String fieldName = field.findFirstToken(TokenTypes.IDENT).getText();
+            if (assignmentsFromConstructor.contains(fieldName) && !assignmentsFromMethods.contains(fieldName)) {
+                log(field, String.format(ERROR_MSG, fieldName));
+            }
+        }
+    }
+
+    /*
+    * Get the field token from an assignation token.
+    * This method handles cases for fields referenced as `this.field` or only `field`
+    * It will get parameters from the method definition to ignore assignations to those parameters
+    */
+    private DetailAST getAssignedField(final DetailAST assignationToken) {
+        final Set<String> scopeParentParameterSet = getParameterSet(scopeParent.findFirstToken(
+            TokenTypes.PARAMETERS));
+        final DetailAST firstChild = assignationToken.getFirstChild();
+        final DetailAST assignationWithDot = firstChild.getType() == TokenTypes.DOT ? firstChild : null;
+
+        if (assignationWithDot != null) {
+            if (assignationWithDot.branchContains(TokenTypes.LITERAL_THIS)) {
+                return assignationWithDot.findFirstToken(TokenTypes.IDENT);
+            }
+        } else {
+            final DetailAST variableNameToken = assignationToken.getFirstChild();
+            // make sure the assignation is not for a method parameter
+            if (!scopeParentParameterSet.contains(variableNameToken.getText())) {
+                return variableNameToken;
+            }
+        }
+
+        return null;
+    }
+
+    /*
+    * Saves a field name to a container depending on the provided type
+    */
+    private void saveField(String fieldName, int scopeParentType) {
```

Comment:
1. Why `int` vs passing `TokenTypes` as the type?
2. What happens if it doesn't match either method or ctor def? I guess I would expect it to throw because this is unexpected.

Reply:
yeah, good point. my bad XD